### PR TITLE
User bookmark count api

### PR DESCRIPTION
**Pull request by sambuddhabasu**

### CODE REVIEW

File: bookie/models/__init__.py [hunk 1]
```diff
@@ -6,6 +6,7 @@
 
 from datetime import datetime
 from datetime import timedelta
+from calendar import monthrange
```
Comment (mitechie) [inline bookie/models/__init__.py]: these should be sorted alphabetically. calendar comes before datetime.

File: bookie/models/__init__.py [hunk 2]
```diff
@@ -469,6 +470,26 @@ def user_dump(username):
             filter(Bmark.username == username).all()
 
     @staticmethod
+    def user_count_bmark(username, start_date=None, end_date=None):
+        """Get a list of user bookmark count
+        """
+        from bookie.models.stats import StatBookmark
+        days = 30
```
Comment (mitechie) [inline bookie/models/__init__.py]: let's make this a module level constant? So up to the top something like STATS_WINDOW = 30

File: bookie/models/__init__.py [hunk 3]
```diff
@@ -469,6 +470,26 @@ def user_dump(username):
...
+        qry = StatBookmark.query.\
```
Comment (mitechie) [inline bookie/models/__init__.py]: this shouldn't talk to the db. It should talk to the StatsMgr. We might have to add a method on there to do the query for us.

File: bookie/models/__init__.py [hunk 4]
```diff
@@ -469,6 +470,26 @@ def user_dump(username):
...
+            filter(StatBookmark.attrib == username)
+        if not start_date:
```
Comment (mitechie) [inline bookie/models/__init__.py]: this gets a little dense to read and follow. Maybe some comments along the way to help out?

File: bookie/models/__init__.py [hunk 5]
```diff
@@ -469,6 +470,26 @@ def user_dump(username):
             filter(Bmark.username == username).all()
 
     @staticmethod
+    def user_count_bmark(username, start_date=None, end_date=None):
```
Comment (mitechie) [inline bookie/models/__init__.py]: this should only be in the stats model.

File: bookie/models/__init__.py (hunk 1 above)
Comment (sambuddhabasu) [inline bookie/models/__init__.py]: changes have been done in https://github.com/sammyshj/Bookie/commit/e0edb6a835d7fe829274f873ffb562390ea18b34

File: bookie/views/api.py [hunk 6]
```diff
@@ -365,6 +365,33 @@ def bmark_recent(request, with_content=False):
     })
 
 
+@view_config(route_name="api_count_bmarks_user", renderer="jsonp")
+@api_auth('api_key', UserMgr.get, anon=False)
+def user_bmark_count(request):
+    """Get a list of bookmark count for the api call"""
```
Comment (mitechie) [inline bookie/views/api.py]: Get the user's daily bookmark total for the given time window. 

Might be a better description?

File: bookie/views/api.py [hunk 7]
```diff
@@ -365,6 +365,33 @@ def bmark_recent(request, with_content=False):
...
+    rdict = request.matchdict
+    params = request.params
+
+    username = 'user_bookmarks_' + request.user.username
```
Comment (mitechie) [inline bookie/views/api.py]: so this should build the string using the constant in the models/stats.py like we do in there already. That way, if we tweak it, there's only one place to make the change.

File: bookie/views/api.py [hunk 8]
```diff
@@ -365,6 +365,33 @@ def bmark_recent(request, with_content=False):
...
+    start_date = params.get('start_date', None)
+    end_date = params.get('end_date', None)
+    bmark_count_list = BmarkMgr.user_count_bmark(
```
Comment (mitechie) [inline bookie/views/api.py]: so then this will turn into the StatsMgr

File: bookie/views/api.py [hunk 9]
```diff
@@ -365,6 +365,33 @@ def bmark_recent(request, with_content=False):
...
+    )
+
+    result_set = []
+    for res in bmark_count_list:
+        return_obj = dict(res)
+
+        result_set.append(return_obj)
+
+    return _api_response(request, {
+        'bookmark_count': result_set
```
Comment (mitechie) [inline bookie/views/api.py]: hmm, the api method is bmarkcount. I'd just make this count and not bookmark_count.

File: bookie/models/stats.py [hunk 10]
```diff
@@ -58,6 +61,17 @@ def get_stat(start, end, *stats):
         return qry.all()
 
     @staticmethod
+    def get_user_stat(username, start_date=None, end_date=None):
+        """Fetch the bookmark count for the user from the stats table"""
+        qry = StatBookmark.query.\
+            filter(StatBookmark.attrib==username)
```
Comment (mitechie) [inline bookie/models/stats.py]: The attrib is stored as  USER_CT.format(username) so when we query the db to get it out, we need to build that matching string back up again. Right now it's just querying for the uesrname without the rest of the prefix.

File: bookie/models/stats.py [hunk 11]
```diff
@@ -58,6 +61,15 @@ def get_stat(start, end, *stats):
         return qry.all()
 
     @staticmethod
+    def get_user_bmark_count(username, start_date, end_date):
+        """Fetch the bookmark count for the user from the stats table"""
+        qry = StatBookmark.query.\
+            filter(StatBookmark.attrib==USER_CT.format(username)).\
+            filter(StatBookmark.tstamp>=start_date).\
+            filter(StatBookmark.tstamp<=end_date)
+        return qry.all()
```
Comment (mitechie) [inline bookie/models/stats.py:73]: we should probably add an order_by on here so that the results always come out in order.

Comment (mitechie) [inline bookie/models/stats.py:73]: http://docs.sqlalchemy.org/en/rel_0_9/orm/tutorial.html#querying for some notes on querying

Comment (sambuddhabasu) [inline bookie/models/stats.py:73]: sure an order_by tstamp would be a good idea

File: bookie/models/stats.py [hunk 12]
```diff
@@ -95,6 +107,24 @@ def count_user_bookmarks(username):
...
+            if not end_date:
+                """If both start_date and end_date are None,
+                end_date will be the current date"""
+                end_date = datetime.utcnow()
+            start_date = end_date - timedelta(days=STATS_WINDOW)
+        elif start_date and not end_date:
+            if start_date.day == 1:
+                """If the starting day is 1, stats of the month is returned"""
+                global STATS_WINDOW
+                STATS_WINDOW = monthrange(start_date.year, start_date.day)[1] - 1
```
Comment (mitechie) [inline bookie/models/stats.py]: Oh, I see. Because you're updating the value here.

Comment (mitechie) [inline bookie/models/stats.py]: I don't think you should update this value. You can use it or use another variable. If someone makes an api call and changes it to 60 days, the next api call will have STATS_WINDOW = the 60 days. It'll break everyone else's api call.

Comment (sambuddhabasu) [inline bookie/models/stats.py]: Changes were made. Thanks for pointing that out

**Comment by mitechie** [comment]
superceded

**Pull request closed without merging**